### PR TITLE
[DCOS-58616] Kafka Kerberos tests

### PR DESCRIPTION
**Pull request by shubhanilBag**

### CODE REVIEW

File: tests/suites/kafka_kerberos/kafka_kerberos_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	. "github.com/mesosphere/kudo-kafka-operator/tests/suites"
+
+	"github.com/mesosphere/kudo-kafka-operator/tests/utils"
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/reporters"
+	. "github.com/onsi/gomega"
+)
+
+var customNamespace = "kerberos-ns"
+var krb5Client = &utils.KDCClient{}
```
Comment (zmalik) [inline tests/suites/kafka_kerberos/kafka_kerberos_test.go]: you can join them in a var block and initialize the krb5Client with variables we already know here: 
```
var (
	customNamespace = "kerberos-ns"
	krb5Client = &utils.KDCClient{
		Namespace: customNamespace,
	}
)
```

File: tests/utils/client.go [hunk 2]
```diff
@@ -59,6 +65,63 @@ func GetKubernetesClient() (*kubernetes.Clientset, error) {
...
+	c.CoreV1().Secrets(namespace).Create(&v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		StringData: map[string]string{
+			data[0]: data[1],
+		},
+	})
+	_, err := c.CoreV1().Secrets(namespace).List(metav1.ListOptions{})
+	// secrets, err := c.CoreV1().Secrets(namespace).List(metav1.ListOptions{})
```
Comment (zmalik) [inline tests/utils/client.go]: let's remove these commented lines

File: tests/utils/kudo.go [hunk 3]
```diff
@@ -104,16 +104,16 @@ func updateInstancesCount(name, namespace string, count int) (string, error) {
...
 	log.Info(fmt.Sprintf("Using kubectl from path: %s", kubectlPath))
 	log.Info(fmt.Sprintf("Installing framework from PATH: %s", resourcesAbsoluteDirectoryPath))
 
 	install_cmd := []string{"kudo", "install", fmt.Sprintf("--instance=%s", name),
 		resourcesAbsoluteDirectoryPath, fmt.Sprintf("--namespace=%s", namespace)}
 
-	if len(params) > 0 {
-		install_cmd = append(install_cmd, "--parameter", params)
+	for key, val := range params {
+		install_cmd = append(install_cmd, "-p", key+"="+val)
```
Comment (zmalik) [inline tests/utils/kudo.go]:
```suggestion
		install_cmd = append(install_cmd, "-p", fmt.Sprintf("%s=%s",key, val))
```

File: tests/suites/kafka_kerberos/kafka_kerberos_test.go (hunk 1 above)
Comment (shubhanilBag) [reply tests/suites/kafka_kerberos/kafka_kerberos_test.go]: PR updated

File: tests/utils/client.go (hunk 2 above)
Comment (shubhanilBag) [reply tests/utils/client.go]: PR updated

File: tests/utils/kudo.go (hunk 3 above)
Comment (shubhanilBag) [reply tests/utils/kudo.go]: PR updated

File: tests/go.mod [hunk 4]
```diff
@@ -7,14 +7,12 @@ require (
 	github.com/elazarl/goproxy v0.0.0-20190421051319-9d40249d3c2f // indirect
 	github.com/elazarl/goproxy/ext v0.0.0-20190421051319-9d40249d3c2f // indirect
 	github.com/jmcvetta/randutil v0.0.0-20150817122601-2bb1b664bcff
+	github.com/kudobuilder/kudo v0.7.3
```
Comment (zmalik) [inline tests/go.mod]: lets not add this dependency also, we are using `go mod edit -require github.com/kudobuilder/kudo@${DS_KUDO_VERSION}` to use the correct version of kudo library in `run.sh`

File: tests/utils/client.go [hunk 5]
```diff
@@ -91,6 +151,25 @@ func (c *KubernetesTestClient) WaitForStatefulSetReadyReplicasCount(name, namesp
...
+	}
+	log.Infof("Found 1 pod with name %s in %s namespace", name, namespace)
+	return true
+}
+
+func (c *KubernetesTestClient) GetPod(name, namespace string) *v1.Pod {
+	pod, err := c.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		log.Warningf("%v", err)
+		return nil
```
Comment (zmalik) [inline tests/utils/client.go]: let's return an `error` here and make the method return `(*v1.Pod, error)`
and handle it `err != nil` where it's using this method.
returning `nil` is not a good pattern unless the type returned is `error` only. 

https://www.digitalocean.com/community/tutorials/handling-errors-in-go

File: tests/utils/kdc.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+// deployPod Use it to deploy the kdc server
+func (k *KDCClient) Deploy() {
+	repoRoot, exists := os.LookupEnv("REPO_ROOT")
+
+	if exists {
+		Create(repoRoot+"/tests/suites/kafka_kerberos/resources/kdc.yaml", k.Namespace)
+		KClient.WaitForPod("kdc", k.Namespace, 240)
+	} else {
+		log.Warningf("Environment variable REPO_ROOT is not set!")
```
Comment (zmalik) [inline tests/utils/kdc.go]: we can use `error` handling here also

File: tests/utils/kudo.go [hunk 7]
```diff
@@ -149,6 +149,19 @@ func (c *KubernetesTestClient) LogObjectsOfKinds(namespace string, components []
 	}
 }
 
+func (c *KubernetesTestClient) PrintLogsOfPod(podName, namespace string) {
```
Comment (zmalik) [inline tests/utils/kudo.go]: this will fail when the pod has more then 1 container, lets pass the container name

File: tests/utils/kudo.go [hunk 8]
```diff
@@ -149,6 +149,19 @@ func (c *KubernetesTestClient) LogObjectsOfKinds(namespace string, components []
...
+	kubectlPath := getKubectlPath()
+	log.Info(fmt.Sprintf("Using kubectl from path: %s", kubectlPath))
+
+	cmd := exec.Command(kubectlPath, "logs", podName, fmt.Sprintf("--namespace=%s", namespace))
+	log.Infoln(fmt.Sprintf("logs %s --namespace=%s", podName, namespace))
+	out, err := cmd.Output()
+	if err != nil {
+		log.Error(string(err.(*exec.ExitError).Stderr))
```
Comment (zmalik) [inline tests/utils/kudo.go]: I think this is unnecessary, casting the error to `ExitError` and getting `Stderr` and can leave out possible `errors` that aren't `ExitError` like if the file isn't an executable or file not found, which `exec.Run` wrap inside generic `Error` type and not `ExitError` type. 

Its ok to just to print the error as it comes, and return it

File: tests/utils/kudo.go [hunk 9]
```diff
@@ -149,6 +149,19 @@ func (c *KubernetesTestClient) LogObjectsOfKinds(namespace string, components []
...
+	}
+	log.Info(fmt.Sprintf(string(out)))
```
Comment (zmalik) [inline tests/utils/kudo.go]: remove `fmt.Sprintf` when there is nothing to format

File: tests/utils/kdc.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+// CreateKeytabSecret Pass it string array of principals and it will create a keytab secret
+func (k *KDCClient) CreateKeytabSecret(principals []string, serviceName string, secretName string) {
+	//
+	command := "printf \"" + strings.Join(principals, "\n") + "\n\" > /kdc/" + serviceName + "-principals.txt;" +
+		"cat /kdc/" + serviceName + "-principals.txt | while read line; do /usr/sbin/kadmin -l add --use-defaults --random-password $line; done;" +
+		"rm /kdc/" + serviceName + ".keytab;" +
+		"cat /kdc/" + serviceName + "-principals.txt | while read line; do /usr/sbin/kadmin -l ext -k /kdc/" + serviceName + ".keytab $line; done;"
+
+	stdout, _ := KClient.ExecInPod(k.Namespace, POD_NAME, CONTAINER_NAME, []string{"/bin/sh", "-c", command})
```
Comment (zmalik) [inline tests/utils/kdc.go]: use the error handling here to see if something went wrong

**Pull request merged**